pkg/sql/rowexec: BenchmarkJoinReaderShortExprs failed

pkg/sql/rowexec.BenchmarkJoinReaderShortExprs [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20289400?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20289400?buildTab=artifacts#./artifacts/microbench/experiment) on master @ [ca0d4076c381609d6487040141c87e36a68b211f](https://github.com/cockroachdb/cockroach/commits/ca0d4076c381609d6487040141c87e36a68b211f):


```
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1/mem=unlimited-32                      	   19669	     58752 ns/op	   0.68 MB/s	   19991 B/op	     103 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=16/mem=unlimited-32                     	    4807	    219821 ns/op	   2.91 MB/s	   39533 B/op	     371 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=128/mem=unlimited-32                    	     852	   1380587 ns/op	   3.71 MB/s	  196653 B/op	    2246 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1024/mem=unlimited-32                   	     100	  10265390 ns/op	   3.99 MB/s	 1463301 B/op	   17853 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=8192/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=8192/mem=unlimited-32                   	      13	  84687708 ns/op	   3.87 MB/s	14628928 B/op	  154422 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited/parallel=true
    joinreader_test.go:1856: expected metadata to contain Metrics but it did not. Err: expected to fetch no more than 1 rows, found 16
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited-32                   	   18037	     64978 ns/op	   2.09 MB/s	   20939 B/op	     108 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=16/mem=unlimited-32                  	    3586	    302648 ns/op	   7.19 MB/s	   54146 B/op	     451 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=128/mem=unlimited-32                 	     555	   2026631 ns/op	   8.59 MB/s	  314093 B/op	    2887 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1024/mem=unlimited-32                	      67	  16081941 ns/op	   8.66 MB/s	 2397382 B/op	   22956 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited/parallel=true
    joinreader_test.go:1856: expected metadata to contain Metrics but it did not. Err: expected to fetch no more than 1 rows, found 32
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited-32                 	   16315	     69396 ns/op	   3.80 MB/s	   22219 B/op	     111 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=16/mem=unlimited-32                	    2684	    401902 ns/op	  10.51 MB/s	   74124 B/op	     499 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=128/mem=unlimited-32               	     396	   2800151 ns/op	  12.07 MB/s	  534174 B/op	    3745 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1024/mem=unlimited-32              	      46	  22189473 ns/op	  12.18 MB/s	 3671650 B/op	   26031 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited/parallel=true
    joinreader_test.go:1856: expected metadata to contain Metrics but it did not. Err: expected to fetch no more than 1 rows, found 64
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited-32                 	   14016	     82031 ns/op	   6.34 MB/s	   24261 B/op	     113 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=16/mem=unlimited-32                	    1809	    610506 ns/op	  13.63 MB/s	  107509 B/op	     531 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=128/mem=unlimited-32               	     247	   4438447 ns/op	  15.00 MB/s	  745068 B/op	    3553 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1024/mem=unlimited-32              	      32	  35312895 ns/op	  15.08 MB/s	 5807700 B/op	   28099 allocs/op
    joinreader_test.go:1874: -- test log scope end --
--- FAIL: BenchmarkJoinReaderShortExprs
FAIL

I250803 23:22:50.796516 1 (gostd) rand.go:268  [-] 1  random seed: -4270515037409731543
test logs left over in: /tmp/logBenchmarkJoinReaderShortExprs3704239520
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkJoinReaderShortExprs.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/sql/rowexec.BenchmarkJoinReaderShortExprs [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20322595?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20322595?buildTab=artifacts#./artifacts/microbench/experiment) on master @ [166d6538fc5bc47b861d81c95e9dbfc19994a1da](https://github.com/cockroachdb/cockroach/commits/166d6538fc5bc47b861d81c95e9dbfc19994a1da):


```
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1/mem=unlimited-32                      	   18916	     59428 ns/op	   0.67 MB/s	   19964 B/op	     103 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=16/mem=unlimited-32                     	    4786	    225747 ns/op	   2.84 MB/s	   39552 B/op	     371 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=128/mem=unlimited-32                    	     816	   1393134 ns/op	   3.68 MB/s	  196622 B/op	    2247 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=1024/mem=unlimited-32                   	      94	  10738653 ns/op	   3.81 MB/s	 1463044 B/op	   17854 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=8192/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetofour/lookuprows=8192/mem=unlimited-32                   	      12	  85773712 ns/op	   3.82 MB/s	15193395 B/op	  155844 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited/parallel=true
    joinreader_test.go:1856: expected metadata to contain Metrics but it did not. Err: expected to fetch no more than 1 rows, found 16
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1/mem=unlimited-32                   	   17841	     66065 ns/op	   2.06 MB/s	   20933 B/op	     108 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=16/mem=unlimited-32                  	    3554	    305222 ns/op	   7.13 MB/s	   54268 B/op	     451 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=128/mem=unlimited-32                 	     542	   1991324 ns/op	   8.74 MB/s	  313478 B/op	    2887 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixteen/lookuprows=1024/mem=unlimited-32                	      66	  16030627 ns/op	   8.69 MB/s	 2621282 B/op	   24223 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited/parallel=true
    joinreader_test.go:1856: expected metadata to contain Metrics but it did not. Err: expected to fetch no more than 1 rows, found 32
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1/mem=unlimited-32                 	   16167	     70291 ns/op	   3.76 MB/s	   22256 B/op	     111 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=16/mem=unlimited-32                	    2739	    402022 ns/op	  10.51 MB/s	   74281 B/op	     499 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=128/mem=unlimited-32               	     390	   2715770 ns/op	  12.44 MB/s	  533640 B/op	    3748 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetothirtytwo/lookuprows=1024/mem=unlimited-32              	      54	  22547613 ns/op	  11.99 MB/s	 3675184 B/op	   26048 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited/parallel=true
    joinreader_test.go:1856: expected metadata to contain Metrics but it did not. Err: expected to fetch no more than 1 rows, found 64
--- FAIL: BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited/parallel=true
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1/mem=unlimited-32                 	   13887	     82754 ns/op	   6.28 MB/s	   24243 B/op	     113 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=16/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=16/mem=unlimited-32                	    1896	    594462 ns/op	  14.00 MB/s	  107386 B/op	     531 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=128/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=128/mem=unlimited-32               	     255	   4214631 ns/op	  15.79 MB/s	  743997 B/op	    3543 allocs/op
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1024/mem=unlimited
BenchmarkJoinReaderShortExprs/reqOrdering=false/matchratio=onetosixtyfour/lookuprows=1024/mem=unlimited-32              	      34	  35157948 ns/op	  15.15 MB/s	 5808078 B/op	   28097 allocs/op
    joinreader_test.go:1874: -- test log scope end --
--- FAIL: BenchmarkJoinReaderShortExprs
FAIL

I250810 11:25:34.855457 1 (gostd) rand.go:268  [-] 1  random seed: -4719381717472276367
test logs left over in: /tmp/logBenchmarkJoinReaderShortExprs578152518
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkJoinReaderShortExprs.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/rowexec/joinreader_test.go
+++ b/pkg/sql/rowexec/joinreader_test.go
@@ -1814,8 +1814,9 @@ func benchmarkJoinReader(b *testing.B, bc JRBenchConfig) {
 
 								spec := execinfrapb.JoinReaderSpec{
 									FetchSpec:           fetchSpec,
-									LookupColumnsAreKey: parallel,
+									LookupColumnsAreKey: parallel && columnDef.matchesPerLookupRow == 1,
 									MaintainOrdering:    reqOrdering,
+									Parallelize:         parallel,
 								}
 								if lookupExpr {
 									// @1 is the column in the input, @2 is the only fetched column.
@@ -2026,9 +2027,10 @@ func BenchmarkJoinReaderLookupStress(b *testing.B) {
 				b.Fatal(err)
 			}
 			spec := execinfrapb.JoinReaderSpec{
-				FetchSpec: fetchSpec,
-				LookupColumnsAreKey:/*parallel=*/ true,
-				MaintainOrdering:/*reqOrdering=*/ false,
+				FetchSpec:           fetchSpec,
+				LookupColumnsAreKey: true,
+				MaintainOrdering:    false,
+				Parallelize:         true,
 			}
 			lookupExprString := "@1 = @2"
 			for i := 0; i < numExprs; i++ {

